File: rofunc/planning/src/robotics-codes-from-scratch-master/python/iLQR_manipulator_object_affordance.py
'''
    Batch iLQR applied to an object affordance planning problem with a planar manipulator, by considering object boundaries.

    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
    Written by Jeremy Maceiras <jeremy.maceiras@idiap.ch>,
    Sylvain Calinon <https://calinon.ch>

    This file is part of RCFS.

    RCFS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 3 as
    published by the Free Software Foundation.

    RCFS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with RCFS. If not, see <http://www.gnu.org/licenses/>.
'''

import copy
import numpy as np
import numpy.matlib
import matplotlib.pyplot as plt
import matplotlib.patches as patches

# Helper functions
# ===============================

# Logarithmic map for R^2 x S^1 manifold
def logmap(f,f0):
    position_error = f[:,:2] - f0[:,:2]
    orientation_error = np.imag(np.log( np.exp( f0[:,-1]*1j ).conj().T * np.exp(f[:,-1]*1j).T )).conj().reshape((-1,1))
    error = np.hstack(( position_error , orientation_error ))
    return error

# Forward kinematics for E-E
def fkin(x):
    global param
    x = x.T
    A = np.tril(np.ones([param["nbVarX"],param["nbVarX"]]))
    f = np.vstack((param["linkLengths"] @ np.cos(A @ x), 
                   param["linkLengths"] @ np.sin(A @ x), 
                   np.mod(np.sum(x,0)+np.pi, 2*np.pi) - np.pi)) #x1,x2,o (orientation as single Euler angle for planar robot)
    return f.T

# Forward Kinematics for all joints
def fkin0(x):
    T = np.tril(np.ones([param["nbVarX"],param["nbVarX"]]))
    T2 = np.tril(np.matlib.repmat(param["linkLengths"],len(x),1))
    f = np.vstack(( 
        T2 @ np.cos(T@x),
        T2 @ np.sin(T@x)
    )).T
    f = np.vstack(( 
        np.zeros(2),
        f
    ))
    return f

def jkin(x):
    global param
    T = np.tril( np.ones((len(x),len(x))) )
    J = np.vstack((
        -np.sin(T@x).T @ np.diag(param["linkLengths"]) @ T,
        np.cos(T@x).T @ np.diag(param["linkLengths"]) @ T,
        np.ones(len(x))
    ))
    return J

# Cost and gradient
def f_reach( x , bounding_boxes = True ):
    global param

    f = logmap(fkin(x),param["mu"])
    J = np.zeros(( len(x) * param["nbVarF"] , len(x) * param["nbVarX"] ))

    for t in range(x.shape[0]):
        f[t,:2] = param["A"][t].T @ f[t,:2] # Object oriented fk
        Jtmp = jkin(x[t])
        Jtmp[:2] = param["A"][t].T @ Jtmp[:2] # Object centered jacobian

        if bounding_boxes:
            for i in range(2):
                if abs(f[t,i]) < param["sizeObj"][i]:
                    f[t,i] = 0
                    Jtmp[i]=0
                else:
                    f[t,i] -=  np.sign(f[t,i]) * param["sizeObj"][i]
        J[ t*param["nbVarF"]:(t+1)*param["nbVarF"] , t*param["nbVarX"]:(t+1)*param["nbVarX"]] = Jtmp
    return f,J

# General param parameters
# ===============================
param = {
    "dt":1e-2,
    "nbData":50,
    "nbIter":50,
    "nbPoints":2,
    "nbVarX":3,
    "nbVarU":3,
    "nbVarF":3,
    "linkLengths":[3,2,1],
    "sizeObj":[.2,.3],
    "tracking_term":1e0,
    "r":1e-6
}

# Task parameters
# ===============================

# Targets
param["mu"] = np.asarray([
    [2,1,-np.pi/6],				# x , y , orientation
    [3,2,-np.pi/3]
])

# Transformation matrices
param["A"] = np.zeros( (param["nbPoints"],2,2) )
for i in range(param["nbPoints"]):
    orn_t = param["mu"][i,-1]
    param["A"][i,:,:] = np.asarray([
        [np.cos(orn_t) , -np.sin(orn_t)],
        [np.sin(orn_t) , np.cos(orn_t)]
    ])

# Regularization matrix
R = np.identity( (param["nbData"]-1) * param["nbVarU"] ) * param["r"]

# Precision matrix
Q = np.identity( param["nbVarF"]  * param["nbPoints"]) * param["tracking_term"]
Qc = copy.deepcopy(Q) # Object affordance constraint matrix

# Constraining the offset of the first and second viapoint to be correlated
Qc[:2,3:5] = -np.identity(2) *1e0
Qc[3:5,:2] = -np.identity(2) *1e0

# System parameters
# ===============================

# Time occurence of viapoints
tl = np.linspace(0,param["nbData"],param["nbPoints"]+1)
tl = np.rint(tl[1:]).astype(np.int64)-1
idx = np.array([ i + np.arange(0,param["nbVarX"],1) for i in (tl* param["nbVarX"])]) 

u = np.zeros( param["nbVarU"] * (param["nbData"]-1) ) # Initial control command
x0 = np.array( [3*np.pi/4 , -np.pi/2 , - np.pi/4] ) # Initial state (in joint space)

# Transfer matrices (for linear system as single integrator)
Su0 = np.vstack([np.zeros((param["nbVarX"], param["nbVarX"]*(param["nbData"]-1))), 
      np.tril(np.kron(np.ones((param["nbData"]-1, param["nbData"]-1)), np.eye(param["nbVarX"])*param["dt"]))]) 
Sx0 = np.kron( np.ones(param["nbData"]) , np.identity(param["nbVarX"]) ).T
Su = Su0[idx.flatten()] # We remove the lines that are out of interest

# Solving iLQR
# ===============================

for i in range( param["nbIter"] ):
    x = Su0 @ u + Sx0 @ x0
    x = x.reshape( (  param["nbData"] , param["nbVarX"]) )

    f, J = f_reach(x[tl])
    fc , Jc = f_reach(x[tl],False)
    du = np.linalg.inv(Su.T @ J.T @ Q @ J @ Su + Su.T @ Jc.T @ Qc @ Jc @ Su + R) @ (-Su.T @ J.T @ Q @ f.flatten() - Su.T @ Jc.T @ Qc @ fc.flatten() - u * param["r"])

    # Perform line search
    alpha = 1
    cost0 = f.flatten() @ Q @ f.flatten() + fc.flatten() @ Qc @ fc.flatten() + np.linalg.norm(u) * param["r"]
    
    while True:
        utmp = u + du * alpha
        xtmp = Su0 @ utmp + Sx0 @ x0
        xtmp = xtmp.reshape( (  param["nbData"] , param["nbVarX"]) )
        ftmp, _ = f_reach(xtmp[tl])
        fctmp, _ = f_reach(xtmp[tl],False)
        cost = ftmp.flatten() @ Q @ ftmp.flatten() + fctmp.flatten() @ Qc @ fctmp.flatten() + np.linalg.norm(utmp) * param["r"]
        
        if cost < cost0 or alpha < 1e-3:
            u = utmp
            print("Iteration {}, cost: {}, alpha: {}".format(i,cost,alpha))
            break

        alpha /=2

# Ploting
# ===============================

plt.figure()
plt.axis("off")
plt.gca().set_aspect('equal', adjustable='box')

# Get points of interest
f = fkin(x)
f00 = fkin0(x[0])
f10 = fkin0(x[tl[0]])
fT0 = fkin0(x[tl[1]])

plt.plot( f00[:,0] , f00[:,1],c='black',linewidth=5,alpha=.2)
plt.plot( f10[:,0] , f10[:,1],c='black',linewidth=5,alpha=.6)
plt.plot( fT0[:,0] , fT0[:,1],c='black',linewidth=5,alpha=1)

plt.plot(f[:,0],f[:,1],c="black",marker="o",markevery=[0]+tl.tolist()) 

# Plot bounding box or via-points
ax = plt.gca()
color_map = ["deepskyblue","darkorange"]
for i in range(param["nbPoints"]):

    rect_origin = param["mu"][i,:2] - param["A"][i]@np.array(param["sizeObj"])
    rect_orn = param["mu"][i,-1]

    rect = patches.Rectangle(rect_origin,param["sizeObj"][0]*2,param["sizeObj"][1]*2,np.degrees(rect_orn),color=color_map[i])
    ax.add_patch(rect)

plt.show()

File: rofunc/planning/src/robotics-codes-from-scratch-master/matlab/iLQR_manipulator_object_affordance.m
%%	  Batch iLQR applied to an object affordance planning problem with a planar manipulator, by considering object boundaries.
%%
%%    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
%%    Written by Sylvain Calinon <https://calinon.ch>
%%
%%    This file is part of RCFS.
%%
%%    RCFS is free software: you can redistribute it and/or modify
%%    it under the terms of the GNU General Public License version 3 as
%%    published by the Free Software Foundation.
%%
%%    RCFS is distributed in the hope that it will be useful,
%%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
%%    GNU General Public License for more details.
%%
%%    You should have received a copy of the GNU General Public License
%%    along with RCFS. If not, see <http://www.gnu.org/licenses/>.

function iLQR_manipulator_object_affordance

%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.dt = 1E-2; %Time step size
param.nbData = 50; %Number of datapoints
param.nbIter = 300; %Maximum number of iterations for iLQR
param.nbPoints = 2; %Number of viapoints
param.nbVarX = 3; %State space dimension (x1,x2,x3)
param.nbVarU = 3; %Control space dimension (dx1,dx2,dx3)
param.nbVarF = 3; %Task space dimension (f1,f2,f3, with f3 as orientation)
param.l = [3; 2; 1]; %Robot links lengths
param.sz = [.2, .3]; %Size of objects
param.q = 1E0; %Tracking weighting term
param.r = 1E-5; %Control weighting term

param.Mu = [[2; 1; -pi/6], [3; 2; -pi/3]]; %Viapoints
for t=1:param.nbPoints
    param.A(:,:,t) = [cos(param.Mu(3,t)), -sin(param.Mu(3,t)); sin(param.Mu(3,t)), cos(param.Mu(3,t))]; %Orientation
end

R = speye(param.nbVarU * (param.nbData-1)) * param.r; %Control weight matrix (at trajectory level)

% Q = kron(eye(param.nbPoints), diag([1E0, 1E0, 0])); %Precision matrix (by removing orientation constraint)
Q = speye(param.nbVarF * param.nbPoints) * param.q; %Precision matrix
Qc = Q; %Object affordance constraint matrix

%Constraining the offset of the first and second viapoint to be correlated
Qc(1:2, 4:5) = -eye(2) * 1E0;
Qc(4:5, 1:2) = -eye(2) * 1E0;

% %Constraining the offset and orientation of the first and second viapoint to be correlated
% Q(1:3, 7:9) = -eye(3) * 1E0;
% Q(7:9, 1:3) = -eye(3) * 1E0;

%Time occurrence of viapoints
tl = linspace(1, param.nbData, param.nbPoints+1);
tl = round(tl(2:end));
idx = (tl - 1) * param.nbVarX + [1:param.nbVarX]';


%% Iterative LQR (iLQR)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
u = zeros(param.nbVarU*(param.nbData-1), 1); %Initial control commands
x0 = [3*pi/4; -pi/2; -pi/4]; %Initial robot pose

%Transfer matrices (for linear system as single integrator)
Su0 = [zeros(param.nbVarX, param.nbVarX*(param.nbData-1)); kron(tril(ones(param.nbData-1)), eye(param.nbVarX)*param.dt)];
Sx0 = kron(ones(param.nbData,1), eye(param.nbVarX));
Su = Su0(idx,:);

for n=1:param.nbIter	
    x = reshape(Su0 * u + Sx0 * x0, param.nbVarX, param.nbData); %System evolution
    [f, J] = f_reach(x(:,tl), param);
    [fc, Jc] = f_reach(x(:,tl), param, 0); %Object affordance
    du = (Su' * J' * Q * J * Su + Su' * Jc' * Qc * Jc * Su + R) \ (-Su' * J' * Q * f(:) -Su' * Jc' * Qc * fc(:) - u * param.r); %Gradient
    
    %Estimate step size with backtracking line search method
    alpha = 1;
    cost0 = f(:)' * Q * f(:) + fc(:)' * Qc * fc(:) + norm(u)^2 * param.r; %u' * R * u
    while 1
        utmp = u + du * alpha;
        xtmp = reshape(Su0 * utmp + Sx0 * x0, param.nbVarX, param.nbData);
        ftmp = f_reach(xtmp(:,tl), param);
        fctmp = f_reach(xtmp(:,tl), param, 0); %Object affordance
        cost = ftmp(:)' * Q * ftmp(:) + fctmp(:)' * Qc * fctmp(:) + norm(utmp)^2 * param.r; %utmp' * R * utmp
        if cost < cost0 || alpha < 1E-3
            break;
        end
        alpha = alpha * 0.5;
    end
    u = u + du * alpha;
    
    if norm(du * alpha) < 1E-2
        break; %Stop iLQR when solution is reached
    end
end
disp(['iLQR converged in ' num2str(n) ' iterations.']);


%% Plot state space
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
msh0 = diag(param.sz) * [-1 -1 1 1 -1; -1 1 1 -1 -1];
for t=1:param.nbPoints
    msh(:,:,t) = param.A(:,:,t) * msh0 + repmat(param.Mu(1:2,t), 1, size(msh0,2));
end

h = figure('position',[10,10,800,800],'color',[1,1,1]); hold on; axis off;
ftmp = fkin0(x(:,1), param);
plot(ftmp(1,:), ftmp(2,:), '-','linewidth',4,'color',[.8 .8 .8]);
ftmp = fkin0(x(:,tl(1)), param);
plot(ftmp(1,:), ftmp(2,:), '-','linewidth',4,'color',[.4 .4 .4]);
ftmp = fkin0(x(:,tl(end)), param);
plot(ftmp(1,:), ftmp(2,:), '-','linewidth',4,'color',[.0 .0 .0]);

colMat = lines(param.nbPoints);
for t=1:param.nbPoints
    patch(msh(1,:,t), msh(2,:,t), min(colMat(t,:)*1.7,1),'linewidth',2,'edgecolor',colMat(t,:)); 
end
ftmp = fkin(x, param); 
plot(ftmp(1,:), ftmp(2,:), 'k-','linewidth',2);
plot(ftmp(1,1), ftmp(2,1), 'k.','markersize',40);
plot(ftmp(1,tl), ftmp(2,tl), 'k.','markersize',30);
axis equal; 

waitfor(h);
end 

%%%%%%%%%%%%%%%%%%%%%%
%Logarithmic map for R^2 x S^1 manifold
function e = logmap(f, f0)
    e(1:2,:) = f0(1:2,:) - f(1:2,:);
    e(3,:) = imag(log(exp(f0(3,:)*1i).' .* exp(f(3,:)*1i)'));
end

%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics (in robot coordinate system)
function f = fkin(x, param)
    L = tril(ones(size(x,1)));
    f = [param.l' * cos(L * x); ...
         param.l' * sin(L * x); ...
         mod(sum(x,1)+pi, 2*pi) - pi]; %f1,f2,f3 (f3 represents orientation as single Euler angle for planar robot)
end

%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics for all robot articulations (in robot coordinate system)
function f = fkin0(x, param)
	L = tril(ones(size(x,1)));
	f = [L * diag(param.l) * cos(L * x), ...
	     L * diag(param.l) * sin(L * x)]'; 
	f = [zeros(2,1), f];
end

%%%%%%%%%%%%%%%%%%%%%%
%Jacobian of forward kinematics function with analytical computation
function J = Jkin(x, param)
    L = tril(ones(size(x,1)));
    J = [-sin(L * x)' * diag(param.l) * L; ...
          cos(L * x)' * diag(param.l) * L; ...
          ones(1, size(x,1))]; %f1,f2,f3 (f3 represents orientation as single Euler angle for planar robot)
end

%%%%%%%%%%%%%%%%%%%%%%
%Cost and gradient for a viapoints reaching task (in object coordinate system)
function [f, J] = f_reach(x, param, bb)
    if nargin<3
        bb = 1; %Bounding boxes considered by default
    end
% 	f = fkin(x, param) - param.Mu; %Error by ignoring manifold
    f = logmap(param.Mu, fkin(x, param)); %Error by considering manifold
    
    J = []; 
    for t=1:size(x,2)
        f(1:2,t) = param.A(:,:,t)' * f(1:2,t); %Object-centered FK
        
        Jtmp = Jkin(x(:,t), param);
        %Jtmp = Jkin_num(x(:,t), param);
        
        Jtmp(1:2,:) = param.A(:,:,t)' * Jtmp(1:2,:); %Object-centered Jacobian
        
        %Bounding boxes 
        if bb==1
            for i=1:2
                if abs(f(i,t)) < param.sz(i)
                    f(i,t) = 0;
                    Jtmp(i,:) = 0;
                else
                    f(i,t) = f(i,t) - sign(f(i,t)) * param.sz(i);
                end
            end
        end
        
        J = blkdiag(J, Jtmp);
    end
end
